dataflux_pytorch/benchmark/checkpointing/multinode/train.py

```diff
@@ -0,0 +1,294 @@
+import argparse
+import os
+import socket
+import time
+import torch
+import statistics
+
+from typing import Generator
+from lightning import Trainer
+from lightning.pytorch.callbacks import ModelCheckpoint
+from lightning.pytorch.demos import (LightningTransformer, Transformer,
+                                     WikiText2)
+import torch.distributed
+from torch.utils.data import DataLoader
+from pathlib import Path
+from google.cloud import storage
+
+from dataflux_core import user_agent
+from dataflux_pytorch.lightning import DatafluxLightningCheckpoint
+from dataflux_pytorch.lightning.gcs_filesystem import GCSDistributedWriter, GCSDistributedReader
+from lightning.pytorch.trainer.states import TrainerFn
+from lightning.pytorch.strategies import FSDPStrategy
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+from lightning.pytorch.strategies.fsdp import _METADATA_FILENAME
+from torch.distributed.checkpoint import save, load
+from torch.nn import Module
+
+
+class DatafluxFSDPStrategy(FSDPStrategy):
+
+    def __init__(self, path, project_name, storage_client, model, enable_multipart, **kwargs):
+        super().__init__(**kwargs)
+        self.writer = GCSDistributedWriter(path, project_name, storage_client)
+        self.reader = GCSDistributedReader(
+            path, project_name, storage_client)
+        self.checkpoint_io = DatafluxLightningCheckpoint(
+            project_name=project_name, storage_client=storage_client, enable_multipart=enable_multipart)
+        self.model = model
+        self.storage_client = storage.Client(project=project_name)
+        user_agent.add_dataflux_user_agent(self.storage_client)
+
+        self.device = torch.device(
+            "cuda" if torch.cuda.is_available() else "cpu")
+
+    def save_checkpoint(self,
+                        checkpoint,
+                        filepath,
+                        storage_options=None) -> None:
+        if storage_options is not None:
+            raise TypeError(
+                "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `FSDPStrategy` does not use the `CheckpointIO`.")
+        path = Path(self.broadcast(filepath))
+        converted_state = {"model": checkpoint.pop("state_dict")}
+        converted_state.update({
+            f"optimizer_{idx}": optim_state
+            for idx, optim_state in enumerate(
+                checkpoint.pop("optimizer_states", []))
+        })
+        save(converted_state, checkpoint_id=path, storage_writer=self.writer)
+
+        if self.global_rank == 0:
+            self.checkpoint_io.save_checkpoint(checkpoint,
+                                               path / _METADATA_FILENAME)
+
+    def get_sharded_state_dict_context(self, module: Module) -> Generator[None, None, None]:
+
+        from torch.distributed.fsdp.api import ShardedOptimStateDictConfig, ShardedStateDictConfig, StateDictType
```

Comment:
Any reason these shouldn't be imported at the top of the script? Similar with other imports inside functions. Keeping all imports at the top helps readability and maintainability.

Reply:
This is 1:1 copy of FSDP source code, and this is something which customers will plug in of their own. Hence my preference is to keep the source code as it is (along with the imports). I do get the comments about readability and maintainability though